File: android/app/src/main/kotlin/com/emfmeter/viewmodel/EMFViewModel.kt
package com.emfmeter.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.emfmeter.data.CalibrationManager
import com.emfmeter.data.EMFProcessor
import com.emfmeter.data.NeedlePhysicsEngine
import com.emfmeter.domain.CalibrationData
import com.emfmeter.domain.DisplayMode
import com.emfmeter.domain.EMFUnit
import com.emfmeter.domain.MeterConfig
import com.emfmeter.domain.ProcessedReading
import com.emfmeter.repository.SettingsRepository
import com.emfmeter.service.AudioService
import com.emfmeter.service.MagnetometerService
import com.emfmeter.util.UnitConverter
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI state for the EMF Meter.
 */
data class EMFUiState(
    val currentReading: ProcessedReading? = null,
    val needlePosition: Float = 0f,
    val displayValue: Float = 0f,
    val displayMode: DisplayMode = DisplayMode.ANALOG,
    val selectedUnit: EMFUnit = EMFUnit.MILLI_GAUSS,
    val soundEnabled: Boolean = true,
    val isCalibrated: Boolean = false,
    val sensorAvailable: Boolean = true,
    val showSettings: Boolean = false,
    val theme: String = "system"
)

@HiltViewModel
class EMFViewModel @Inject constructor(
    private val magnetometerService: MagnetometerService,
    private val audioService: AudioService,
    private val settingsRepository: SettingsRepository
) : ViewModel() {

    private val calibrationManager = CalibrationManager()
    private val emfProcessor = EMFProcessor(calibrationManager)
    private val needlePhysics = NeedlePhysicsEngine()

    private val _uiState = MutableStateFlow(EMFUiState())
    val uiState: StateFlow<EMFUiState> = _uiState.asStateFlow()

    private var needleUpdateJob: Job? = null

    init {
        loadSettings()
        observeReadings()
        startNeedleAnimation()

        _uiState.update {
            it.copy(sensorAvailable = magnetometerService.isAvailable)
        }
    }

    private fun loadSettings() {
        viewModelScope.launch {
            // Load all saved settings
            val unit = settingsRepository.selectedUnit.first()
            val soundEnabled = settingsRepository.soundEnabled.first()
            val displayMode = settingsRepository.displayMode.first()
            val theme = settingsRepository.theme.first()
            val calibration = settingsRepository.calibrationData.first()

            if (calibration.isCalibrated) {
                calibrationManager.restore(calibration)
            }

            _uiState.update { state ->
                state.copy(
                    selectedUnit = unit,
                    soundEnabled = soundEnabled,
                    displayMode = displayMode,
                    theme = theme,
                    isCalibrated = calibration.isCalibrated
                )
            }

            audioService.setEnabled(soundEnabled)
        }
    }

    private fun observeReadings() {
        viewModelScope.launch {
            magnetometerService.readings.collect { reading ->
                val processed = emfProcessor.process(reading)

                _uiState.update { state ->
                    val displayValue = UnitConverter.convert(
                        processed.magnitude,
                        EMFUnit.MICRO_TESLA,
                        state.selectedUnit
                    )

                    state.copy(
                        currentReading = processed,
                        displayValue = displayValue
                    )
                }

                // Play click sound if in analog mode with sound enabled
                val state = _uiState.value
                if (state.soundEnabled && state.displayMode == DisplayMode.ANALOG) {
                    audioService.playClickIfNeeded(processed.normalizedValue)
                }
            }
        }
    }

    private fun startNeedleAnimation() {
        needleUpdateJob = viewModelScope.launch {
            val frameTime = 1000L / MeterConfig.DISPLAY_REFRESH_HZ
            while (isActive) {
                val targetPosition = _uiState.value.currentReading?.normalizedValue ?: 0f
                val newPosition = needlePhysics.update(
                    targetPosition,
                    deltaTime = 1f / MeterConfig.DISPLAY_REFRESH_HZ
                )

                _uiState.update { it.copy(needlePosition = newPosition) }
                delay(frameTime)
            }
        }
    }

    fun setDisplayMode(mode: DisplayMode) {
        _uiState.update { it.copy(displayMode = mode) }
        viewModelScope.launch {
            settingsRepository.saveDisplayMode(mode)
        }
    }

    fun setUnit(unit: EMFUnit) {
        _uiState.update { state ->
            val newDisplayValue = state.currentReading?.let {
                UnitConverter.convert(it.magnitude, EMFUnit.MICRO_TESLA, unit)
            } ?: 0f

            state.copy(
                selectedUnit = unit,
                displayValue = newDisplayValue
            )
        }
        viewModelScope.launch {
            settingsRepository.saveUnit(unit)
        }
    }

    fun toggleSound() {
        val newState = !_uiState.value.soundEnabled
        _uiState.update { it.copy(soundEnabled = newState) }
        audioService.setEnabled(newState)
        viewModelScope.launch {
            settingsRepository.saveSoundEnabled(newState)
        }
    }

    fun calibrate() {
        val currentReading = _uiState.value.currentReading?.rawReading ?: return
        val calibration = calibrationManager.calibrate(currentReading)

        _uiState.update { it.copy(isCalibrated = true) }

        viewModelScope.launch {
            settingsRepository.saveCalibration(calibration)
        }
    }

    fun resetCalibration() {
        calibrationManager.reset()
        _uiState.update { it.copy(isCalibrated = false) }
        viewModelScope.launch {
            settingsRepository.clearCalibration()
        }
    }

    fun setTheme(theme: String) {
        _uiState.update { it.copy(theme = theme) }
        viewModelScope.launch {
            settingsRepository.saveTheme(theme)
        }
    }

    fun showSettings(show: Boolean) {
        _uiState.update { it.copy(showSettings = show) }
    }

    fun onStart() {
        magnetometerService.start()
    }

    fun onStop() {
        magnetometerService.stop()
    }

    override fun onCleared() {
        super.onCleared()
        needleUpdateJob?.cancel()
        magnetometerService.stop()
        audioService.release()
    }
}

File: ios/EMFMeter/Sources/ViewModels/EMFViewModel.swift
import SwiftUI
import Combine

// MARK: - UserDefaults Keys

private enum EMFViewModelKeys {
    static let unit = "selectedUnit"
    static let soundEnabled = "soundEnabled"
    static let displayMode = "displayMode"
    static let theme = "theme"
    static let calibrationX = "calibrationX"
    static let calibrationY = "calibrationY"
    static let calibrationZ = "calibrationZ"
    static let calibrationTimestamp = "calibrationTimestamp"
    static let hasLaunchedBefore = "hasLaunchedBefore"
}

/// ViewModel for the EMF Meter.
@MainActor
class EMFViewModel: ObservableObject {
    // Services - using type-erased wrapper for protocol
    private let magnetometerService: AnyMagnetometerService
    private let audioService = AudioService()
    private let needlePhysics = NeedlePhysicsEngine()

    // Calibration
    private var calibrationData: CalibrationData = .none
    private var needsAutoCalibration: Bool = false

    // Display link for animation
    private var displayLink: CADisplayLink?
    private var cancellables = Set<AnyCancellable>()

    // Published state
    @Published var needlePosition: Float = 0
    @Published var displayValue: Float = 0
    @Published var displayMode: DisplayMode = .analog
    @Published var selectedUnit: EMFUnit = .microTesla
    @Published var soundEnabled: Bool = true
    @Published var isCalibrated: Bool = false
    @Published var sensorAvailable: Bool = true
    @Published var showSettings: Bool = false
    @Published var theme: String = "system"

    /// Current raw reading (exposed for recording service)
    @Published private(set) var currentReading: EMFReading?

    // Current processed reading (internal)
    private var processedReading: ProcessedReading?

    /// Initialize with automatic service selection (mock on simulator, real on device).
    /// Also supports UI testing launch arguments for different scenarios.
    convenience init() {
        let service = MagnetometerServiceFactory.create()
        self.init(magnetometerService: AnyMagnetometerService(service))
    }

    /// Initialize with a specific magnetometer service (for testing/customization).
    init<T: MagnetometerServiceProtocol>(magnetometerService: T) {
        self.magnetometerService = AnyMagnetometerService(magnetometerService)
        loadSettings()
        setupBindings()
        setupDisplayLink()
        sensorAvailable = self.magnetometerService.isAvailable
    }

    /// Initialize with a type-erased service directly.
    init(magnetometerService: AnyMagnetometerService) {
        self.magnetometerService = magnetometerService
        loadSettings()
        setupBindings()
        setupDisplayLink()
        sensorAvailable = magnetometerService.isAvailable
    }

    private func loadSettings() {
        let defaults = UserDefaults.standard

        // Load unit
        if let unitString = defaults.string(forKey: EMFViewModelKeys.unit),
           let unit = EMFUnit(rawValue: unitString) {
            selectedUnit = unit
        }

        // Load sound
        soundEnabled = defaults.object(forKey: EMFViewModelKeys.soundEnabled) as? Bool ?? true
        audioService.isEnabled = soundEnabled

        // Load display mode
        if let modeString = defaults.string(forKey: EMFViewModelKeys.displayMode),
           let mode = DisplayMode(rawValue: modeString) {
            displayMode = mode
        }

        // Load theme
        theme = defaults.string(forKey: EMFViewModelKeys.theme) ?? "system"

        // Load calibration
        let calibrationTimestamp = defaults.double(forKey: EMFViewModelKeys.calibrationTimestamp)
        if calibrationTimestamp > 0 {
            calibrationData = CalibrationData(
                offsetX: defaults.float(forKey: EMFViewModelKeys.calibrationX),
                offsetY: defaults.float(forKey: EMFViewModelKeys.calibrationY),
                offsetZ: defaults.float(forKey: EMFViewModelKeys.calibrationZ),
                timestamp: calibrationTimestamp
            )
            isCalibrated = true
        }

        // Check if this is the first launch - if so, auto-calibrate once we get a reading
        let hasLaunchedBefore = defaults.bool(forKey: EMFViewModelKeys.hasLaunchedBefore)
        if !hasLaunchedBefore {
            needsAutoCalibration = true
            defaults.set(true, forKey: EMFViewModelKeys.hasLaunchedBefore)
        }
    }

    private func setupBindings() {
        magnetometerService.$currentReading
            .compactMap { $0 }
            .sink { [weak self] reading in
                self?.processReading(reading)
            }
            .store(in: &cancellables)
    }

    private func setupDisplayLink() {
        displayLink = CADisplayLink(target: self, selector: #selector(updateNeedle))
        displayLink?.preferredFrameRateRange = CAFrameRateRange(
            minimum: 30,
            maximum: 60,
            preferred: 60
        )
        displayLink?.add(to: .main, forMode: .common)
    }

    @objc private func updateNeedle() {
        guard let reading = processedReading else { return }

        let newPosition = needlePhysics.update(
            targetPosition: reading.normalizedValue,
            deltaTime: 1.0 / Float(MeterConfig.displayRefreshHz)
        )
        needlePosition = newPosition

        if soundEnabled {
            audioService.playClickIfNeeded(normalizedValue: reading.normalizedValue)
        }
    }

    private func processReading(_ reading: EMFReading) {
        // Auto-calibrate on first launch once we have a valid reading
        if needsAutoCalibration {
            calibrationData = CalibrationData.from(reading: reading)
            isCalibrated = true
            needsAutoCalibration = false

            // Persist the auto-calibration
            let defaults = UserDefaults.standard
            defaults.set(calibrationData.offsetX, forKey: EMFViewModelKeys.calibrationX)
            defaults.set(calibrationData.offsetY, forKey: EMFViewModelKeys.calibrationY)
            defaults.set(calibrationData.offsetZ, forKey: EMFViewModelKeys.calibrationZ)
            defaults.set(calibrationData.timestamp, forKey: EMFViewModelKeys.calibrationTimestamp)
        }

        let calibrated = calibrationData.apply(to: reading)
        let magnitude = calibrated.magnitude
        let normalized = min(max(magnitude / MeterConfig.maxValueUT, 0), 1)

        currentReading = calibrated  // Expose calibrated reading for recording/oscilloscope
        processedReading = ProcessedReading(
            rawReading: reading,
            calibratedReading: calibrated,
            magnitude: magnitude,
            normalizedValue: normalized
        )

        displayValue = UnitConverter.convert(magnitude, from: .microTesla, to: selectedUnit)
    }

    // MARK: - Public Actions

    func setDisplayMode(_ mode: DisplayMode) {
        displayMode = mode
        UserDefaults.standard.set(mode.rawValue, forKey: EMFViewModelKeys.displayMode)
    }

    func setUnit(_ unit: EMFUnit) {
        selectedUnit = unit
        if let reading = processedReading {
            displayValue = UnitConverter.convert(reading.magnitude, from: .microTesla, to: unit)
        }
        UserDefaults.standard.set(unit.rawValue, forKey: EMFViewModelKeys.unit)
    }

    func toggleSound() {
        audioService.playSwitch()
        soundEnabled.toggle()
        audioService.isEnabled = soundEnabled
        UserDefaults.standard.set(soundEnabled, forKey: EMFViewModelKeys.soundEnabled)
    }

    func calibrate() {
        audioService.playPushButton()
        guard let reading = magnetometerService.currentReading else { return }

        calibrationData = CalibrationData.from(reading: reading)
        isCalibrated = true

        let defaults = UserDefaults.standard
        defaults.set(calibrationData.offsetX, forKey: EMFViewModelKeys.calibrationX)
        defaults.set(calibrationData.offsetY, forKey: EMFViewModelKeys.calibrationY)
        defaults.set(calibrationData.offsetZ, forKey: EMFViewModelKeys.calibrationZ)
        defaults.set(calibrationData.timestamp, forKey: EMFViewModelKeys.calibrationTimestamp)
    }

    func resetCalibration() {
        calibrationData = .none
        isCalibrated = false

        let defaults = UserDefaults.standard
        defaults.removeObject(forKey: EMFViewModelKeys.calibrationX)
        defaults.removeObject(forKey: EMFViewModelKeys.calibrationY)
        defaults.removeObject(forKey: EMFViewModelKeys.calibrationZ)
        defaults.removeObject(forKey: EMFViewModelKeys.calibrationTimestamp)
    }

    func setTheme(_ theme: String) {
        self.theme = theme
        UserDefaults.standard.set(theme, forKey: EMFViewModelKeys.theme)
    }

    func start() {
        magnetometerService.start()
    }

    func stop() {
        magnetometerService.stop()
    }

    deinit {
        displayLink?.invalidate()
    }
}

// Extension for Float UserDefaults
extension UserDefaults {
    func float(forKey key: String) -> Float {
        Float(double(forKey: key))
    }
}
